cluster.c

```diff
@@ -86,29 +100,70 @@ RRStatus ShardGroupDeserialize(const char *buf, size_t buf_len, ShardGroup *sg)
     str[buf_len] = '\0';
     char *s = str;
 
-    /* Find and null terminate header */
-    char *nl = strchr(str, '\n');
-    if (!nl) goto error;
+    memset(sg, 0, sizeof(*sg));
+
+    /* Find and null terminate shard group id */
+    char *nl = strchr(s, '\n');
+    if (!nl) {
+        goto error;
+    }
     *nl = '\0';
 
-    memset(sg, 0, sizeof(*sg));
+    /* Validate shard group id */
+    if (nl - s > RAFT_DBID_LEN) {
+        goto error;
+    }
+
+    memcpy(sg->id, s, nl - s);
+    s = nl + 1;
+
+    /* Find and null terminate slot_ranges_num and nodes_num header */
+    nl = strchr(s, '\n');
+    if (!nl) {
+        goto error;
+    }
+    *nl = '\0';
 
-    if (sscanf(s, "%u:%u:%u:%u", &sg->id, &sg->start_slot, &sg->end_slot, &sg->nodes_num) != 4)
+    if (sscanf(s, "%u:%u", &sg->slot_ranges_num, &sg->nodes_num) != 2) {
         goto error;
+    }
     s = nl + 1;
 
+    /* Read in slot ranges */
+    sg->slot_ranges = RedisModule_Alloc(sizeof(ShardGroupSlotRange) * sg->slot_ranges_num);
```

Comment:
Suggest to use RedisModule_Calloc and let it handle the potential integer overflow (this is unsanitized external input).